docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+Pure functions have many advantages but when using JAX they are often essential. For example, most code requires compilation using `jax.jit` to be fast and once you created a Module you probably want to optimize its parameters using `jax.grad`. However, these APIs expect a pure function and don't work on stateful bound `Module` instances directly. Moreover, pure functions allow for flexible interoperability with other libraries.
+
+### Cloning
+
+To make this approach work reliably we need well-defined cloning behavior. Rather than relying on a complex nested cloning procedure like Python's `deepcopy`, Flax enforces that a `Module` is exactly defined by its construction arguments. Therefore cloning a Module reduces to calling the constructor with its original construction arguments. Because `Module` acts as an immutable dataclass, the construction arguments are mapped 1:1 to instance attributes.
+
+### Bind
+
+Sometimes it's useful to have a bound, top-level Module without having to wrap the code in a function. For example: to interact with a Module inside a Jupyter notebook. The [bind](https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=bind#flax.linen.Module.bind) method returns a bound clone with an unlimited lifetime. The downside of this is that you cannot combine it with JAX transformations or integrate it into a vanilla JAX codebase that expects stateless code.
+
+
+## Setup
+
+The `setup` method is often used like the constructor hook (`__init__`) in normal Python classes. However, for more advanced use cases it’s good to realize that it is not quite the same as a constructor.
+
+`setup` is only called after a Module becomes bound. Normally, this is not an issue because most Modules are bound (almost) immediately (as part of `init` and `apply`). Inside `setup`, sub-modules become bound when they are assigned to an attribute. nside an `nn.compact` decorated method, sub-modules are bound immediately when constructed. As explained in the previous section, top-level Modules are never bound and thus setup is not called when they are constructed. This means that you cannot access attributes assigned in setup from an unbound, top-level module.
+
+The `setup` method is not called immediately after the `Module` becomes bound but only when you interact with the `Module` instance (e.g.: call a method or access an attribute). This should not impact the behavior of a `Module` but the lazy execution does sometimes affect log statements and stack traces during debugging. The section on functionalization will explain why we need `setup` to be lazy in the first place.
+
+
+# Functionalization
```

Comment:
This whole section: should it really be part of this document, or is it maybe a better fit for the lifted transformations design note?

It seems not so much related to "Linen module lifecycle" to me, but rather about "pure functions that can directly be transformed vs. functions inside bound modules that must take extra care to transform state that is implicitly handled by Flax"

Reply:
I do think this is part of the lifecycle because functionalization affects multiple design decisions (lazy setup, cloning, ...) and effectively inside a lifted transformation there is a mini/recurisve lifecycle.